### PR TITLE
Query: Simplify member access on conditional null check constructing …

### LINKED ISSUE #20711: OrderBy throws InvalidOperationException on complex type when Select contains null check
### Steps to reproduce

``` C#
//Entities namespace
 public partial class SubRegion
 {
    public int Id { get; set; }
    public int RegionId { get; set; }
    public string Name { get; set; }
    public Region Region { get; set; }
 }

public partial class Region
{
    public int Id { get; set; }
    public string Name { get; set; }
    public ICollection<SubRegion> SubRegion { get; set; }
}

//Models namespace
public class SubRegion 
{
    public int Id { get; set; }
    public string Name { get; set; }
    public Option Region { get; set; }
}

public class Option 
{
    public int Id { get; set; }
    public string Name { get; set; }
}
```

OrderBy throws InvalidOperationException if called on a complex type when Select contains null check like so:

```csharp
 var subRegions = _context.SubRegion
                .Select(
                s => new Models.SubRegion
                {
                  Id = s.Id,
                  Name = s.Name,
                  Region = (s.Region == null)
                        ? null
                        : new Models.Option
                        {
                          Id = s.Region.Id,
                          Name = s.Region.Name
                        }
                })
                .OrderBy(s => s.Region.Name);

```
Error Message:

```csharp
System.InvalidOperationException: The LINQ expression 'DbSet<SubRegion>
    .Join(
        outer: DbSet<Region>, 
        inner: s => EF.Property<Nullable<int>>(s, "RegionId"), 
        outerKeySelector: r => EF.Property<Nullable<int>>(r, "Id"), 
        innerKeySelector: (o, i) => new TransparentIdentifier<SubRegion, Region>(
            Outer = o, 
            Inner = i
        ))
    .OrderBy(s => EF.Property<Nullable<int>>(s.Inner, "Id") == null ? null : new Option{ 
        Id = s.Inner.Id, 
        Name = s.Inner.Name 
    }
    .Name)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync().
```

If you remove the null check it works:

```csharp
var subRegions = _context.SubRegion
                .Select(
                s => new Models.SubRegion
                {
                  Id = s.Id,
                  Name = s.Name,
                  Region = new Models.Option
                  {
                    Id = s.Region.Id,
                    Name = s.Region.Name
                  }
                })
                .OrderBy(s => s.Region.Name);
```

I discovered this using the Automapper ProjectTo function, which adds null checks by default, and initially filed an issue [there](https://github.com/AutoMapper/AutoMapper/issues/3385).
 
### Further technical details

EF Core version: 3.1.3
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Windows 10 Enterprise
IDE: Visual Studio 2019 16.5.4

### PR DESCRIPTION
**Pull request by smitpatel**
…anonymous type

Resolves #20711

### CODE REVIEW

File: src/EFCore/Query/Internal/QueryOptimizingExpressionVisitor.cs [hunk 1]
```diff
@@ -28,6 +28,61 @@ public class QueryOptimizingExpressionVisitor : ExpressionVisitor
...
+                    || binaryTest.NodeType == ExpressionType.NotEqual)
+                // Exclude HasValue/Value over Nullable<> as they return non-null type and we don't have equivalent for it for null part
+                && !(conditionalExpression.Type.IsNullableValueType()
+                    && (visitedMemberExpression.Member.Name == nameof(Nullable<int>.HasValue)
+                        || visitedMemberExpression.Member.Name == nameof(Nullable<int>.Value))))
+            {
+                var isLeftNullConstant = IsNullConstant(binaryTest.Left);
+                var isRightNullConstant = IsNullConstant(binaryTest.Right);
+
+                if (isLeftNullConstant != isRightNullConstant
```
Comment (maumar) [inline src/EFCore/Query/Internal/QueryOptimizingExpressionVisitor.cs:56]: we could also optimize the other way around, i.e. (a != null ? null : new { Member = b, ... }).Member a != null ? null : b, or even generalize member access over conditional: pushdown member access to iftrue and iffalse, and if member access is over null constant, just return the constant. but prolly less risky to do targeted optimization

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 2]
```diff
@@ -8375,6 +8375,159 @@ private SqlServerTestStore CreateDatabase21807()
 
         #endregion
 
+        #region Issue20711
+
+        [ConditionalFact]
+        public virtual void Simplify_member_access_on_null_conditional_check()
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:8381]: why not use complex navs or gears of war? you didn't want to create the dto? model seems trivial

File: src/EFCore/Query/Internal/QueryOptimizingExpressionVisitor.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/QueryOptimizingExpressionVisitor.cs:56]: The main goal of this is to make sure null removal can run and shorten it. Anything else while likely be work but not sure if it works in all cases. Like member access over null can resolve to null constant but what if the type is non-nullable then it is going to null ref. Which raises the question (about the query you gave example of), why write query so badly that it is causing null ref. The specific case being fixed is only targets scenarios which are kind of generated by AM or OData - "the null checks".

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:8381]: Will move. While looking at issue, I thought the DTOs were essential to repro.

**Pull request merged**